### PR TITLE
Fix grpc bugs

### PR DESCRIPTION
**Pull request by gongweibao**
Fix https://github.com/PaddlePaddle/Paddle/issues/7520#event-1423729037
Fix https://github.com/grpc/grpc/issues/13983

### CODE REVIEW

File: paddle/operators/detail/grpc_client.cc [hunk 1]
```diff
@@ -110,22 +109,25 @@ bool RPCClient::Proceed() {
...
   }
-  req_count_--;
 
   GPR_ASSERT(ok);
   PADDLE_ENFORCE(tag);
 
   // TODO(gongwb): add more retries.
   ClientBase* c = static_cast<ClientBase*>(tag);
   if (!c->status_.ok()) {
+    LOG(ERROR) << "proc param error:" << c->var_h_.String();
```
Comment (typhoonzero) [inline paddle/operators/detail/grpc_client.cc]: Only log one time for the error.

File: paddle/operators/send_op.cc [hunk 2]
```diff
@@ -48,7 +48,10 @@ class SendOp : public framework::OperatorBase {
       client_.AsyncGetVariable(epmap[i], ctx, scope, outs[i]);
     }
 
-    client_.wait();
+    if (!client_.wait()) {
+      LOG(ERROR) << "send op exit";
+      exit(1);
```
Comment (typhoonzero) [inline paddle/operators/send_op.cc]: Do **not** use `exit` in operators, use `PADDLE_ENFORCE`

File: python/paddle/v2/fluid/tests/book_distribute/test_dist_image_classification.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
+# Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved
```
Comment (typhoonzero) [inline python/paddle/v2/fluid/tests/book_distribute/test_dist_image_classification.py]: Can you please split the fix and the book dist sample in two PRs?

File: paddle/operators/detail/grpc_server.cc [hunk 4]
```diff
@@ -184,15 +183,20 @@ void AsyncGRPCServer::HandleRequest(bool wait, grpc::ServerCompletionQueue* cq,
       break;
     }
 
+    assert(tag);
```
Comment (typhoonzero) [inline paddle/operators/detail/grpc_server.cc]: Use `PADDLE_ENFORCE`

File: paddle/operators/recv_op.cc [hunk 5]
```diff
@@ -100,6 +100,8 @@ class RecvOp : public framework::OperatorBase {
       // TODO(gognwb): simply this loop.
       // Get from multiple trainers, we don't care about order in which
       // the gradient arrives, just add suffix 0~n then average the gradient.
+      VLOG(4) << "param_count:" << param_count
```
Comment (typhoonzero) [inline paddle/operators/recv_op.cc]: Reduce `VLOG` appearances.

File: paddle/operators/detail/grpc_server.cc [hunk 6]
```diff
@@ -28,12 +28,15 @@ class RequestBase {
  public:
   explicit RequestBase(sendrecv::SendRecvService::AsyncService* service,
                        grpc::ServerCompletionQueue* cq)
-      : service_(service), cq_(cq), status_(PROCESS) {}
+      : service_(service), cq_(cq), status_(PROCESS) {
+    assert(cq_);
```
Comment (typhoonzero) [inline paddle/operators/detail/grpc_server.cc]: `PADDLE_ENFORCE`

File: paddle/operators/detail/grpc_server.cc [hunk 7]
```diff
@@ -184,15 +183,20 @@ void AsyncGRPCServer::HandleRequest(bool wait, grpc::ServerCompletionQueue* cq,
...
+    // reference:
+    // https://github.com/tensorflow/tensorflow/issues/5596
+    // https://groups.google.com/forum/#!topic/grpc-io/xftlRy-IQwM
     if (!ok) {
-      VLOG(4) << cq_name << " recv no regular event";
+      LOG(WARNING) << cq_name << " recv no regular event:argument name"
+                   << base->GetReqName();
+      // FIXME(gongwb): delete the old one?
       TryToRegisterNewOne();
-      delete base;
```
Comment (typhoonzero) [inline paddle/operators/detail/grpc_server.cc]: No other places to release this memory then.

File: paddle/operators/detail/grpc_server.cc [hunk 8]
```diff
@@ -184,15 +183,20 @@ void AsyncGRPCServer::HandleRequest(bool wait, grpc::ServerCompletionQueue* cq,
...
 
     RequestBase* base = (RequestBase*)tag;
+    // reference:
+    // https://github.com/tensorflow/tensorflow/issues/5596
+    // https://groups.google.com/forum/#!topic/grpc-io/xftlRy-IQwM
     if (!ok) {
-      VLOG(4) << cq_name << " recv no regular event";
+      LOG(WARNING) << cq_name << " recv no regular event:argument name"
+                   << base->GetReqName();
+      // FIXME(gongwb): delete the old one?
```
Comment (typhoonzero) [inline paddle/operators/detail/grpc_server.cc]: This comment does not make things clear.

File: paddle/operators/send_op.cc [hunk 9]
```diff
@@ -48,7 +48,10 @@ class SendOp : public framework::OperatorBase {
       client_.AsyncGetVariable(epmap[i], ctx, scope, outs[i]);
     }
 
-    client_.wait();
+    if (!client_.wait()) {
+      LOG(ERROR) << "send op exit";
```
Comment (typhoonzero) [inline paddle/operators/send_op.cc]: This log is too simple.

File: paddle/operators/detail/grpc_server.cc (hunk 6 above)
Comment (gongweibao) [reply paddle/operators/detail/grpc_server.cc]: Done.

File: paddle/operators/detail/grpc_server.cc (hunk 4 above)
Comment (gongweibao) [reply paddle/operators/detail/grpc_server.cc]: Done.

File: paddle/operators/detail/grpc_server.cc (hunk 8 above)
Comment (gongweibao) [reply paddle/operators/detail/grpc_server.cc]: Can't get more context when `ok != true`.

File: paddle/operators/detail/grpc_server.cc (hunk 7 above)
Comment (gongweibao) [reply paddle/operators/detail/grpc_server.cc]: So add FixME.

File: paddle/operators/recv_op.cc (hunk 5 above)
Comment (gongweibao) [reply paddle/operators/recv_op.cc]: Done.

File: paddle/operators/send_op.cc (hunk 9 above)
Comment (gongweibao) [reply paddle/operators/send_op.cc]: Detail logs has logged in `Wait()`.

File: paddle/operators/send_op.cc (hunk 2 above)
Comment (gongweibao) [reply paddle/operators/send_op.cc]: Done.

File: paddle/operators/detail/grpc_client.cc (hunk 1 above)
Comment (gongweibao) [reply paddle/operators/detail/grpc_client.cc]: Done.

File: paddle/operators/detail/grpc_server.cc (hunk 7 above)
Comment (gongweibao) [reply paddle/operators/detail/grpc_server.cc]: Done.

Comment (gongweibao) [reply paddle/operators/detail/grpc_server.cc]: I'm not sure it's a grpc bug or it's our application bug.
When delete base often, I met an [error](https://github.com/grpc/grpc/issues/13983)

Comment (gongweibao) [reply paddle/operators/detail/grpc_server.cc]: So add `FixMe` here.

File: python/paddle/v2/fluid/tests/book_distribute/test_dist_image_classification.py (hunk 3 above)
Comment (gongweibao) [reply python/paddle/v2/fluid/tests/book_distribute/test_dist_image_classification.py]: Done.

**Pull request merged**